src/event/cloudevent.ts

```diff
@@ -51,7 +51,7 @@ export class CloudEvent implements CloudEventV1 {
    * @param {object} event the event properties
    * @param {boolean?} strict whether to perform event validation when creating the object - default: true
    */
-  constructor(event: CloudEventV1 | CloudEventV1Attributes, strict = true) {
+  constructor(event: CloudEventV1<T> | CloudEventV1Attributes<T>, strict = true) {
```

Comment:
I like this simplification. It does, however, require me to cast on line 65 and 68 - where this was not required before. I suppose this is because they are required in `CloudEvent` but not in a `Partial`, and assigning `undefined` makes the type checker unhappy.

```typescript
    this.type = properties.type as string;
    delete (properties as any).type;

    this.source = properties.source as string;
    delete (properties as any).source;
```

Reply:
It actually makes sense that you need to cast these. For example in your "loose validation" unit test, you initialize this class like:

```typescript
const ce = new CloudEvent({}, false);
```

So in this case `properties.type` and `properties.source` would be `undefined`. The type checker knows this is possible so it is forcing you to cast.

Maybe I misunderstood the intent here. Should clients be allowed do initialize a `CloudEvent` the way you do in the "loose validation" UT or do you want that to be a type error?

Maybe this is all that you need:

```typescript
constructor(event: CloudEventV1Attributes<T>, strict = true) {
```

That should give you:

```typescript
// OK
const ce = new CloudEvent({
  type: "",
  source: "",
}, false); 

// type error! the type and source properties are required
const ce = new CloudEvent({}, false); 

// OK - because the client added a cast to satisfy the type checker
const ce = new CloudEvent({} as CloudEventV1Attributes<unknown>, false); 
```

The other option is to make the type check less rigid when the `strict` flag is passed in. Something like this:

```
// overload one, no strict flag so required props are required
constructor(event: CloudEventV1Attributes<T>)
// overload two, basically anything goes
constructor(event: Partial<CloudEventV1Attributes<T>>, strict: boolean)
// actual implementation
constructor(event: CloudEventV1Attributes<T>, strict = true) {
```

That should give you:

```typescript
// OK
const ce = new CloudEvent({
  type: "",
  source: "",
}); 

// type error! the type and source properties are required
const ce = new CloudEvent({}); 

// OK
const ce = new CloudEvent({}, false); 

// OK - ideally this would be a type error but, AFAIK you can't express that in TS
const ce = new CloudEvent({}, true); 
```

I think either option makes sense. The difference is very subtle.

Reply:
Noted above - it's working as expected.